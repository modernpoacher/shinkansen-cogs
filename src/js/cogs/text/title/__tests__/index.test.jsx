import React from 'react'
import classnames from 'classnames'

import {
  snapshotOf
} from 'react-component-snapshot'

import {
  getInstanceFromContainerElement
} from 'react-component-instance'

import '@testing-library/jest-dom'

import {
  render
} from '@testing-library/react'

import Super from '#cogs/components/title'
import Title from '#cogs/cogs/text/title'

jest.mock('classnames', () => jest.fn().mockReturnValue('MOCK CLASSNAME'))

describe('#cogs/cogs/text/title', () => {
  describe('<Title />', () => {
    describe('With required props', () => {
      it('renders', () => {
        const {
          container: {
            firstElementChild: title
          }
        } = render(
          <Title
            id='MOCK ID'
          />
        )

        expect(title)
          .toBeNull()
      })

      describe('Always', () => {
        it('invokes `getClassName`', () => {
          const getClassNameSpy = jest.spyOn(Title.prototype, 'getClassName')

          render(
            <Title
              id='MOCK ID'
            />
          )

          expect(getClassNameSpy)
            .not.toHaveBeenCalled()
        })
      })

      /**
       *  Element is null
       */
      it('matches the snapshot', () => {
        const {
          container: {
            firstElementChild: title
          }
        } = render(
          <Title
            id='MOCK ID'
          />
        )

        expect(snapshotOf(title))
          .toMatchSnapshot()
      })
    })

    describe('With additional props', () => {
      it('renders', () => {
        const {
          container: {
            firstElementChild: title
          }
        } = render(
          <Title
            id='MOCK ID'
            title='MOCK TITLE'
            required
            disabled
            readOnly
          />
        )

        expect(title)
          .toBeInstanceOf(HTMLLabelElement)
      })

      describe('Always', () => {
        it('invokes `getClassName`', () => {
          const getClassNameSpy = jest.spyOn(Title.prototype, 'getClassName')

          render(
            <Title
              id='MOCK ID'
              title='MOCK TITLE'
              required
              disabled
              readOnly
            />
          )

          expect(getClassNameSpy)
            .toHaveBeenCalled()
        })
      })

      it('matches the snapshot', () => {
        const {
          container: {
            firstElementChild: title
          }
        } = render(
          <Title
            id='MOCK ID'
            title='MOCK TITLE'
            required
            disabled
            readOnly
          />
        )

        expect(snapshotOf(title))
          .toMatchSnapshot()
      })
    })

    describe('`getClassName()`', () => {
      it('invokes `classnames`', () => {
        /**
         *  Ensure `super.getClassName()` returns a value
         */
        const getClassNameSpy = jest.spyOn(Super.prototype, 'getClassName').mockReturnValue('MOCK CLASSNAME')

        const {
          container
        } = render(
          <Title
            id='MOCK ID'
          />
        )

        const instance = getInstanceFromContainerElement(container)

        /**
         *  Ensure it is reset after render
         */
        classnames.mockClear()

        /**
         *  Ensure it is reset after render
         */
        getClassNameSpy.mockClear()

        instance.getClassName()

        expect(classnames)
          .toHaveBeenCalledWith('MOCK CLASSNAME', 'text')
      })
    })
  })
})
