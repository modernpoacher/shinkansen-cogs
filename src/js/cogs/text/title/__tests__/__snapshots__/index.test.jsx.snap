// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#cogs/cogs/text/title <Title /> With additional props matches the snapshot 1`] = `
<label
  className="MOCK CLASSNAME"
  htmlFor="MOCK ID"
>
  <TextContent>
    <span
      className="text-content"
    >
      MOCK TITLE
    </span>
  </TextContent>
  <Required>
    <span
      className="is-required"
    >
      *
    </span>
  </Required>
  <Disabled>
    <span
      className="is-disabled"
    >
      *
    </span>
  </Disabled>
  <ReadOnly>
    <span
      className="is-readonly"
    >
      *
    </span>
  </ReadOnly>
</label>
`;

exports[`#cogs/cogs/text/title <Title /> With required props matches the snapshot 1`] = `null`;
